python/ray/util/client/dataclient.py

```diff
@@ -73,66 +78,157 @@ def _start_datathread(self) -> threading.Thread:
+        Process responses from the data servicer.
+        """
+        if response.req_id == 0:
+            # This is not being waited for.
+            logger.debug(f"Got unawaited response {response}")
+            return
+        if response.req_id in self.asyncio_waiting_data:
+            try:
+                # NOTE: calling self.asyncio_waiting_data.pop() results
+                # in the destructor of ClientObjectRef running, which
+                # calls ReleaseObject(). So self.asyncio_waiting_data
+                # is accessed without holding self.lock. Holding the
+                # lock shouldn't be necessary either.
+                callback = self.asyncio_waiting_data.pop(response.req_id)
+                if callback:
+                    callback(response)
+            except Exception:
+                logger.exception("Callback error:")
             with self.lock:
-                self._in_shutdown = True
-                self._last_exception = e
+                # Update outstanding requests
+                if response.req_id in self.outstanding_requests:
+                    del self.outstanding_requests[response.req_id]
+                    # Acknowledge response
+                    self._acknowledge(response.req_id)
+        else:
+            # Blocking response. Populate ready data and wake up waiting
+            # threads.
+            with self.cv:
+                self.ready_data[response.req_id] = response
                 self.cv.notify_all()
 
-                callbacks = self.asyncio_waiting_data.values()
-                self.asyncio_waiting_data = {}
+    def _can_reconnect(self, e: grpc.RpcError) -> bool:
+        """
+        Processes RPC errors that occur while reading from data stream.
+        Returns True if the error can be recovered from, False otherwise.
+        """
+        if not self.client_worker._can_reconnect(e):
+            logger.info("Unrecoverable error in data channel.")
+            logger.debug(e)
+            self._last_exception = e
+            return False
+        logger.debug("Recoverable error in data channel.")
+        logger.debug(e)
+        return True
+
+    def _shutdown(self) -> None:
+        """
+        Shutdown the data channel
+        """
+        with self.lock:
+            self._in_shutdown = True
+            self.cv.notify_all()
+
+            callbacks = self.asyncio_waiting_data.values()
+            self.asyncio_waiting_data = {}
 
+        if self._last_exception:
             # Abort async requests with the error.
-            err = ConnectionError("Failed during this or a previous request. "
-                                  f"Exception that broke the connection: {e}")
+            err = ConnectionError(
+                "Failed during this or a previous request. Exception that "
+                f"broke the connection: {self._last_exception}")
             for callback in callbacks:
-                callback(err)
+                if callback:
```

Comment:
Assert `callbacks` are empty when `self._last_exception` is None, and add a comment for the reason (all pending requests have finished)?

Reply:
I don't think that's guaranteed to be the case. For example, the user might ctrl-c the session while there are outstanding async requests, which would gracefully shut down the channel (insert a None into the request_queue)

Reply:
I would think that will not be a graceful shutdown. Anyway this seems very subtle, and we would need to run the `callbacks` regardless of if `self._last_exception` is not `None`.

Reply:
Sounds good, will send an error in both cases. In case this case becomes important eventually, the general flow in the situation I'm thinking of would be:
* User Ctrl-C's
* The main thread (thread that the driver program is running on) receives the error, and begins to exit while doing on async get
* The client worker's close method is called
* The data_client's close method is called
* `None` is dumped into request_queue
* The data_main for loop exits without error (sentinel `None` hit) and goes through the , possible with some outstanding async callbacks

Basically, a Ctrl-C to the driver thread will still lead to a graceful shutdown of the data_main thread